### ISSUE TITLE
Allow specifying custom TokenCache when creating new credentials

### ISSUE THREAD

**Issue by ratzrattillo** [Azure/azure-sdk-for-go#19445]
### Feature Request
I would like to be able to specify a custom TokenCache when creating new credentials using one of the `azidentity.NewXYZCredential` Methods.

The reason is, that restarting an application always requires a new login. I want to reuse the tokens from the previous login.

How could this potentially be achieved?

Imagine having a custom TokenCache writing credentials to a file:

```go
type TokenCache struct {
	file string
}

func (t *TokenCache) Replace(cache cache.Unmarshaler, key string) {
	data, err := os.ReadFile(t.file)
	if err != nil {
		log.Println(err)
	}
	err = cache.Unmarshal(data)
	if err != nil {
		log.Println(err)
	}
}

func (t *TokenCache) Export(cache cache.Marshaler, key string) {
	data, err := cache.Marshal()
	if err != nil {
		log.Println(err)
	}
	err = os.WriteFile(t.file, data, 0600)
	if err != nil {
		log.Println(err)
	}
}
```

In my program, i would like to instruct the method `azidentity.NewDeviceCodeCredential(options)` to use this TokenCache like:

```go
cacheAccessor := &TokenCache{"serialized_cache.json"}
cred, err = azidentity.NewDeviceCodeCredential(options, cacheAccessor)
if err != nil {
    ...
}
```

In the implementation of `azidentity.NewDeviceCodeCredential` only smaller changes should be required.
These changes might include allowing one or multiple optional parameters like e.g. the cacheAccessor in the method's signature. If such an option is set, the way the client is built should be done with e.g. `public.WithCache(accessor)`:

```go
// NewDeviceCodeCredential creates a DeviceCodeCredential. Pass nil to accept default options.
func NewDeviceCodeCredential(options *DeviceCodeCredentialOptions, accessor cache.ExportReplace) (*DeviceCodeCredential, error) {
	cp := DeviceCodeCredentialOptions{}
	if options != nil {
		cp = *options
	}
	cp.init()
	if !validTenantID(cp.TenantID) {
		return nil, errors.New(tenantIDValidationErr)
	}
	authorityHost, err := setAuthorityHost(cp.Cloud)
	if err != nil {
		return nil, err
	}
	c, err := public.New(cp.ClientID,
		public.WithAuthority(runtime.JoinPaths(authorityHost, cp.TenantID)),
		public.WithHTTPClient(newPipelineAdapter(&cp.ClientOptions)),
		public.WithCache(accessor),
	)
	if err != nil {
		return nil, err
	}
	return &DeviceCodeCredential{userPrompt: cp.UserPrompt, client: c}, nil
}
```

The goal is to reuse the RefreshToken and AccessTokens without having to log in after shutting down an application, because this information can be pulled again from the custom Cache.
So in this case, after successfully logging in once, the following popup: `To sign in, use a web browser to open the page https://microsoft.com/devicelogin and enter the code XYZABCXYZ to authenticate.` should not appear, until the RefreshToken is revoked.

Please let me know your feedback! Did i oversee options that already make this possible?

**Comment by chlowell** [comment]
Thanks for the detailed issue! This isn't something we support today but we do plan to add optional persistent caching so applications don't need to interactively authenticate every time they run (#16643). We've done that in the Azure SDK for other languages, and plan to follow their design in Go. One difference from your proposal is that the other Azure SDKs keep the implementation and details of the persistent cache internal. Applications set options for the Azure SDK's implementation instead of writing their own implementation. There are only a couple options: a name for the cache, so applications can keep their caches separate from others, and a bool to allow unencrypted storage when encryption isn't possible (the SDK always attempts encryption first). Do you think that would work for your application?

**Comment by ratzrattillo** [comment]
That is exactly what i am looking for. Persistent and either encrypted or unencrypted storage would be great. Most important would be that it survives reboots without issues.
It would be amazing to have this functionality as fast as possible. Some organizations only allow interactive oauth codeflows and this is currently a pain in the ass, if you want to not sign in again and again.
Is there already a timeframe for the implementation of this feature that we can look forward to?

**Comment by chlowell** [comment]
It's on our list of features to release by next April. I don't yet know when it will first appear in beta, but if you subscribe to #16643, you'll know when that happens. I'm marking this issue as addressed because we already have #16643 tracking this feature, but please feel free to comment on that issue or continue discussion here.
